server/src-lib/Hasura/GraphQL/Execute/Inline.hs

```diff
@@ -0,0 +1,176 @@
+
+> query {
+>   users {
+>     id
+>     ...userFields
+>   }
+> }
+>
+> fragment userFields on User {
+>   name
+>   favoriteColor
+> }
+
+the fragment inliner will convert it to this:
+
+> query {
+>   users {
+>     id
+>     ... on User {
+>       name
+>       favoriteColor
+>     }
+>   }
+> }
+
+This is a straightforward and mechanical transformation, but it simplifies
+further processing, since we catch unbound fragments and recursive fragment
+definitions early in the pipeline, so parsing does not have to worry about it.
+In that sense, fragment inlining is similar to the variable resolution pass
+performed by "Hasura.GraphQL.Execute.Resolve", but for fragment definitions
+rather than variables. -}
+module Hasura.GraphQL.Execute.Inline
+  ( inlineSelectionSet
+  ) where
+
+import           Hasura.Prelude
+
+import qualified Data.HashMap.Strict.Extended  as Map
+import qualified Data.HashSet                  as Set
+import qualified Data.List                     as L
+import qualified Data.Text                     as T
+
+import           Control.Lens
+import           Language.GraphQL.Draft.Syntax
+
+import           Hasura.RQL.Types.Error
+import           Hasura.Server.Utils
+import           Hasura.SQL.Types
+
+-- | Internal bookkeeping used during inlining.
+data InlineEnv = InlineEnv
+  { _ieFragmentDefinitions :: HashMap Name FragmentDefinition
+  -- ^ All known fragment definitions.
+  , _ieFragmentStack       :: [Name]
+  -- ^ Fragments we’re currently inlining higher up in the call stack, used to
+  -- detect fragment cycles.
+  }
+
+-- | Internal bookkeeping used during inlining.
+newtype InlineState var = InlineState
+  { _isFragmentCache :: HashMap Name (InlineFragment NoFragments var)
+  -- ^ A cache of fragment definitions we’ve already inlined, so we don’t need
+  -- to inline them again.
+  }
+
+$(makeLensesFor [("_ieFragmentStack", "ieFragmentStack")] ''InlineEnv)
+$(makeLenses ''InlineState)
+
+type MonadInline var m =
+  ( MonadError QErr m
+  , MonadReader InlineEnv m
+  , MonadState (InlineState var) m
+  )
+
+-- | Inlines all fragment spreads in a 'SelectionSet'; see the module
+-- documentation for "Hasura.GraphQL.Execute.Inline" for details.
+inlineSelectionSet
+  :: (MonadError QErr m, Foldable t)
+  => t FragmentDefinition
+  -> SelectionSet FragmentSpread var
+  -> m (SelectionSet NoFragments var)
+inlineSelectionSet fragmentDefinitions selectionSet = do
+  let fragmentDefinitionMap = Map.groupOnNE _fdName fragmentDefinitions
+  uniqueFragmentDefinitions <- flip Map.traverseWithKey fragmentDefinitionMap
+    \fragmentName fragmentDefinitions' ->
+      case fragmentDefinitions' of
+        a :| [] -> return a
+        _       -> throw400 ParseFailed $ "multiple definitions for fragment " <>> fragmentName
+  let usedFragmentNames = Set.fromList $ fragmentsInSelectionSet selectionSet
+      definedFragmentNames = Set.fromList $ Map.keys uniqueFragmentDefinitions
+      isFragmentValidationEnabled = False
```

Comment:
We should leave a TODO / a pretty explicit comment.

Reply:
I'll add the comment!

Reply:
Done